File: tbb/chain/randmat.cc
/* randmat: random number generator
 *
 * input:
 *   nrows, ncols: number of rows and columns
 *   s: random number generation seed
 *
 * output:
 *   matrix: random nrows x ncols integer matrix
 */
#include <cstdlib>

#include <vector>

#include "tbb/parallel_for.h"
#include "tbb/blocked_range.h"

using namespace std;
using namespace tbb;

typedef tbb::blocked_range<size_t> range;

void randmat(int nrows, int ncols, int s, vector<vector<int> >* matrix) {
  srand(s);
  parallel_for(
      range(0, nrows),
      [&](range r) {
        for (size_t i = r.begin(); i != r.end(); ++i) {
          for (int j = 0; j < ncols; j++) {
            (*matrix)[i][j] = rand() % 1000;
          }
        }
      });
}

File: go/chain/src/all/randmat.go
/*
 * randmat: random number generation
 *
 * input:
 *   nrows, ncols: the number of rows and columns
 *   s: the seed
 *
 * output:
 *   martix: a nrows x ncols integer matrix
 *
 */
package all

import (
  "math/rand"
)

func randvec(vector []int, n int, done chan bool) {
  for i := 0; i < n; i++ {
    vector[i] = rand.Int() % 1000;
  }
  done <- true;
}

func Randmat(nrows, ncols, s int) [][]int {
  var matrix [][]int;
  matrix = make([][]int, nrows);
  for i := 0; i < nrows; i++ {
    matrix[i] = make([]int, ncols);
  }
  rand.Seed(int64(s));
  done := make(chan bool);
  for i := 0; i < nrows; i++ {
    go randvec(matrix[i], ncols, done);
  }
  for i := 0; i < nrows; i++ {
    <-done;
  }
  return matrix;
}

